subxt/src/rpc/lightclient/client.rs

```diff
@@ -0,0 +1,376 @@
+struct LightClientInner {
+    /// Smoldot light client implementation that leverages the `AsyncStdTcpWebSocket`.
+    ///
+    /// Note: `AsyncStdTcpWebSocket` is not wasm compatible.
+    client: smoldot_light::Client<AsyncStdTcpWebSocket>,
+    /// The ID of the chain used to identify the chain protocol (ie. substrate).
+    ///
+    /// Note: A single chain is supported for a client. This aligns with the subxt's
+    /// vision of the Client.
+    chain_id: ChainId,
+    /// Communicate with the backend task.
+    to_backend: mpsc::Sender<BackendMessage>,
+    /// Atomic used to generate unique IDs.
+    id_provider: AtomicU64,
+}
+
+impl LightClientInner {
+    /// Generate the next unique ID used to populate the Json RPC request.
+    ///
+    /// This is unique to identify the sender of the request.
+    fn next_id(&mut self) -> String {
+        let id = self.id_provider.fetch_add(1, Ordering::AcqRel);
+        id.to_string()
+    }
+
+    /// Register a RPC method request.
+    ///
+    /// Returns a channel that produces only one item, which is the result of the method.
+    ///
+    /// The result is a raw jsonrpc string similar to:
+    ///
+    /// ```bash
+    /// {"jsonrpc":"2.0","id":"1","result":"my result object"}
+    /// ```
+    ///
+    /// # Note
+    ///
+    /// Registering the request must happen before submitting the request in order
+    /// for the background task to provide a response.
+    async fn register_request(
+        &self,
+        id: String,
+    ) -> Result<oneshot::Receiver<Box<RawValue>>, LightClientError> {
+        let (sender, receiver) = oneshot::channel();
+
+        self.to_backend
+            .send(BackendMessage::Request { id, sender })
+            .await
+            .map_err(|_| LightClientError::BackgroundClosed)?;
+
+        Ok(receiver)
+    }
+
+    /// Register a RPC subscription request.
+    ///
+    /// Returns a channel that produces the items of the subscription.
+    ///
+    /// The JsonRPC subscription is generated as follows:
+    /// - Make a plain RPC method request which returns the subscription ID, in the result field:
+    ///
+    /// ```bash
+    /// {"jsonrpc":"2.0","id":"1","result":"0"}
+    /// ```
+    ///
+    /// - Register with the provided ID to the notifications of the subscription. Notifications look like:
+    ///
+    /// ```bash
+    /// {"jsonrpc":"2.0","method":"author_extrinsicUpdate","params":{"subscription":"0","result":"Dropped"}}
+    /// ```
+    ///
+    /// # Note
+    ///
+    /// The notification messages are buffered internally to ensure that users will receive all
+    /// messages in the following case:
+    ///
+    /// * T0. [`Self::register_request()`].
+    /// * T1. submit a plain RPC method request.
+    /// * T2. the subscription produces a notification. (T2 happens before the user calls this method)
+    /// * T3. user parses the subscription ID from (T1) and calls [`Self::register_subscription`].
```

Comment:
ok, I see why the cache is needed then. A little weird I would expect to register_subscription to occur before then reading the subscription notifications.